README.md
**Note: This project is not maintained anymore. croquemort does not exist anymore. See [hydra](https://github.com/datagouv/hydra) for a udata crawler**

# uData-croquemort

This plugin provides integration between uData and Croquemort link checker.

## Compatibility

**udata-croquemort** requires Python 2.7+ and [uData][].

## Installation

Install [uData][].

Remain in the same virtual environment (for Python) and use the same version of npm (for JS).

Install **udata-croquemort**:

```shell
pip install udata-croquemort
```

Modify your local configuration file of **udata** (typically, `udata.cfg`) as following:

```python
PLUGINS = ['croquemort']
LINKCHECKING_DEFAULT_LINKCHECKER = 'croquemort'
CROQUEMORT_URL = 'http://localhost:8000'
CROQUEMORT_NB_RETRY = 10
CROQUEMORT_DELAY = 1
```

[uData]: https://github.com/opendatateam/udata


CHANGELOG.md
# Changelog

## Current (in progress)

- Nothing yet

## 2.0.1 (2020-06-17)

- Expose new HTTP headers related to content (`content-type`, `content-length`, `charset`, `content-md5`, `content-disposition`) [#133](https://github.com/opendatateam/udata-croquemort/pull/133)

## 2.0.0 (2020-03-11)

- Migrate to python3 🐍 [#93](https://github.com/opendatateam/udata-croquemort/pull/93)

## 1.0.4 (2019-05-20)

- Do not check FTP(S) urls [#86](https://github.com/opendatateam/udata-croquemort/pull/86)

## 1.0.3 (2018-07-30)

- Handle croquemort response w/o updated attr [#31](https://github.com/opendatateam/udata-croquemort/pull/31)

## 1.0.2 (2018-03-13)

- Switch to pytest

## 1.0.1 (2017-10-20)

- Fix setup.py version

## 1.0.0 (2017-10-20)

- Stable version

## 0.0.1 (2017-10-03)

- Initial release


udata_croquemort/tests.py
import json
from datetime import datetime

import pytest
import requests

# import this first or hell will break loose
from udata.core.dataset.factories import DatasetFactory, ResourceFactory
from udata.settings import Testing
from udata.utils import faker

from .checker import CroquemortLinkChecker

CROQUEMORT_TEST_URL = 'http://check.test'
CHECK_ONE_URL = '{0}/check/one'.format(CROQUEMORT_TEST_URL)
METADATA_URL = '{0}/url'.format(CROQUEMORT_TEST_URL)


def metadata_factory(url, data=None):
    """Base for a mocked Croquemort HTTP response"""
    response = {
        'etag': '',
        'checked-url': url,
        'content-length': faker.pyint(),
        'content-disposition': '',
        'content-md5': faker.md5(),
        'content-location': '',
        'expires': faker.iso8601(),
        'final-status-code': 200,
        'updated': faker.iso8601(),
        'last-modified': faker.iso8601(),
        'content-encoding': 'gzip',
        'content-type': faker.mime_type()
    }
    if data:
        response.update(data)
    return json.dumps(response)


def mock_url_check(httpretty, url, data=None, status=200):
    url_hash = faker.md5()
    httpretty.register_uri(httpretty.POST, CHECK_ONE_URL,
                           body=json.dumps({'url-hash': url_hash}),
                           content_type='application/json')
    check_url = '/'.join((METADATA_URL, url_hash))
    httpretty.register_uri(httpretty.GET, check_url,
                           body=metadata_factory(url, data),
                           content_type='application/json',
                           status=status)


def exception_factory(exception):
    def callback(request, uri, headers):
        raise exception
    return callback


class CheckUrlSettings(Testing):
    CROQUEMORT_URL = CROQUEMORT_TEST_URL
    CROQUEMORT_NB_RETRY = 3
    CROQUEMORT_DELAY = 1


class UdataCroquemortTest:
    settings = CheckUrlSettings

    @pytest.fixture(autouse=True)
    def setup(self, app):
        self.resource = ResourceFactory()
        self.dataset = DatasetFactory(resources=[self.resource])
        self.checker = CroquemortLinkChecker()

    def test_returned_metadata(self, httpretty):
        url = self.resource.url
        test_cases = [
            {'status': 200, 'available': True},
            {'status': 301, 'available': True},
            {'status': 404, 'available': False},
            {'status': 500, 'available': False},
            {'status': 503, 'available': False},
        ]
        metadata = {
            'content-type': 'text/html',
            'content-length': '2512124',
            'charset': 'utf-8'
        }
        for test_case in test_cases:
            metadata['final-status-code'] = test_case['status']
            mock_url_check(httpretty, url, metadata)
            res = self.checker.check(self.resource)
            assert res['check:url'] == url
            assert res['check:status'] == test_case['status']
            assert res['check:available'] == test_case['available']
            assert isinstance(res['check:date'], datetime)
            assert res['check:headers:content-type'] == 'text/html'
            assert res['check:headers:content-length'] == '2512124'
            assert res['check:headers:charset'] == 'utf-8'

    def test_returned_metadata_w_missing_updated(self, httpretty):
        url = self.resource.url
        metadata = {
            'content-type': 'text/html',
            'final-status-code': 200,
            'available': True
        }
        mock_url_check(httpretty, url, metadata)
        res = self.checker.check(self.resource)
        assert res['check:status'] == 200

    def test_post_request(self, httpretty):
        url = self.resource.url
        url_hash = faker.md5()
        httpretty.register_uri(httpretty.POST, CHECK_ONE_URL,
                               body=json.dumps({'url-hash': url_hash}),
                               content_type='application/json')
        check_url = '/'.join((METADATA_URL, url_hash))
        httpretty.register_uri(httpretty.GET, check_url,
                               body=metadata_factory(url),
                               content_type='application/json',
                               status=200)
        self.checker.check(self.resource)
        assert len(httpretty.core.httpretty.latest_requests)
        post_request = httpretty.core.httpretty.latest_requests[0]
        assert json.loads(post_request.body) == {
            'url': self.resource.url,
            'group': self.dataset.slug
        }

    def test_delayed_url(self, httpretty):
        url = faker.uri()
        mock_url_check(httpretty, url, status=404)
        res = self.checker.check(self.resource)
        assert res is None

    def test_timeout(self, httpretty):
        exception = requests.Timeout('Request timed out')
        httpretty.register_uri(httpretty.POST, CHECK_ONE_URL,
                               body=exception_factory(exception))
        res = self.checker.check(self.resource)
        assert res is None

    def test_connection_error(self, httpretty):
        exception = requests.ConnectionError('Unable to connect')
        httpretty.register_uri(httpretty.POST, CHECK_ONE_URL,
                               body=exception_factory(exception))
        res = self.checker.check(self.resource)
        assert res is None

    def test_json_error_check_one(self, httpretty):
        httpretty.register_uri(httpretty.POST, CHECK_ONE_URL,
                               body='<strong>not json</strong>',
                               content_type='text/html')
        res = self.checker.check(self.resource)
        assert res is None

    def test_json_error_check_url(self, httpretty):
        url_hash = faker.md5()
        httpretty.register_uri(httpretty.POST, CHECK_ONE_URL,
                               body=json.dumps({'url-hash': url_hash}),
                               content_type='application/json')
        check_url = '/'.join((METADATA_URL, url_hash))
        httpretty.register_uri(httpretty.GET, check_url,
                               body='<strong>not json</strong>',
                               content_type='text/html')
        res = self.checker.check(self.resource)
        assert res is None

    def test_retry(self, httpretty):
        '''Test the `is_pending` logic from utils.check_url'''
        url = self.resource.url
        url_hash = faker.md5()
        httpretty.register_uri(httpretty.POST, CHECK_ONE_URL,
                               body=json.dumps({'url-hash': url_hash}),
                               content_type='application/json')
        check_url = '/'.join((METADATA_URL, url_hash))

        def make_response(status, body=None):
            return httpretty.Response(body=body or metadata_factory(url),
                                      content_type='application/json',
                                      status=status)

        httpretty.register_uri(httpretty.GET, check_url,
                               responses=[
                                   make_response(500),
                                   make_response(404, body=''),
                                   make_response(200)
                               ])
        res = self.checker.check(self.resource)
        assert res['check:status'], 200

    def test_ftp_url(self, httpretty):
        resource = ResourceFactory(url='Ftp://etalab.gouv.fr')
        DatasetFactory(resources=[resource])
        res = self.checker.check(resource)
        assert res is None


class NoCroquemortConfiguredTest:
    def test_croquemort_not_configured(self, app):
        dataset = DatasetFactory(visible=True)
        checker = CroquemortLinkChecker()

        assert checker.check(dataset.resources[0]) is None


tasks.py
import os

from invoke import task, call

ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__)))

TO_CLEAN = ['build', 'dist', '**/*.pyc', 'reports']


def color(code):
    '''A simple ANSI color wrapper factory'''
    return lambda t: '\033[{0}{1}\033[0;m'.format(code, t)


green = color('1;32m')
red = color('1;31m')
blue = color('1;30m')
cyan = color('1;36m')
purple = color('1;35m')
white = color('1;39m')


def header(text):
    '''Display an header'''
    print(' '.join((blue('>>'), cyan(text))))


def info(text, *args, **kwargs):
    '''Display informations'''
    text = text.format(*args, **kwargs)
    print(' '.join((purple('>>>'), text)))


def success(text):
    '''Display a success message'''
    print(' '.join((green('>>'), white(text))))


def error(text):
    '''Display an error message'''
    print(red('✘ {0}'.format(text)))


@task
def clean(ctx):
    '''Cleanup all build artifacts'''
    header(clean.__doc__)
    with ctx.cd(ROOT):
        for pattern in TO_CLEAN:
            info('Removing {0}', pattern)
            ctx.run('rm -rf {0}'.format(pattern))


@task
def test(ctx, report=False):
    '''Run tests suite'''
    cmd = 'pytest -v'
    if report:
        cmd = ' '.join((cmd, '--junitxml=reports/tests.xml'))
    with ctx.cd(ROOT):
        ctx.run(cmd, pty=True)


@task
def cover(ctx, html=False):
    '''Run tests suite with coverage'''
    cmd = 'pytest --cov udata_croquemort --cov-report term'
    if html:
        cmd = ' '.join((cmd, '--cov-report html:reports/cover'))
    with ctx.cd(ROOT):
        ctx.run(cmd, pty=True)


@task
def qa(ctx):
    '''Run a quality report'''
    header(qa.__doc__)
    with ctx.cd(ROOT):
        info('Python Static Analysis')
        flake8_results = ctx.run('flake8 udata_croquemort', pty=True, warn=True)
        if flake8_results.failed:
            error('There is some lints to fix')
        else:
            success('No lint to fix')
    if flake8_results.failed:
        error('Quality check failed')
        exit(flake8_results.return_code)
    success('Quality check OK')


@task
def dist(ctx, buildno=None):
    '''Package for distribution'''
    header('Building a distribuable package')
    cmd = ['python setup.py']
    if buildno:
        cmd.append('egg_info -b {0}'.format(buildno))
    cmd.append('bdist_wheel')
    with ctx.cd(ROOT):
        ctx.run(' '.join(cmd), pty=True)
    success('Distribution is available in dist directory')


@task(clean, qa, call(test, report=True), dist, default=True)
def default(ctx):
    '''Perform quality report, tests and packaging'''
    pass


udata_croquemort/utils.py
from http import client as httplib
import json
import logging
import time

import requests

from flask import current_app

log = logging.getLogger(__name__)

# See http://docs.python-requests.org/en/latest/user/advanced/#timeouts
# We are waiting 3 sec for the connexion and 9 for the response.
TIMEOUT = (3.1, 9.1)

DEFAULT_DELAY = 5
DEFAULT_RETRY = 10
CONNECTION_ERROR_MSG = 'Unable to reach the URL checker'

ERROR_LOG_MSG = 'Unable to connect to croquemort'
TIMEOUT_LOG_MSG = 'Timeout connecting to Croquemort'


class UnreachableLinkChecker(Exception):
    pass


def is_pending(response):
    if not response:
        return True
    if response.status_code == httplib.NOT_FOUND:
        return True
    try:
        return 'final-status-code' not in response.json()
    except ValueError:
        return True


def check_url(url, group=None):
    """Check the given URL against a Croquemort server.

    Return a tuple: (error, response).
    """
    croquemort_url = current_app.config.get('CROQUEMORT_URL')
    if croquemort_url is None:
        raise UnreachableLinkChecker('Croquemort server not configured')
    croquemort_check_url = '{url}/check/one'.format(url=croquemort_url)
    delay = current_app.config.get('CROQUEMORT_DELAY', DEFAULT_DELAY)
    retry = current_app.config.get('CROQUEMORT_NB_RETRY', DEFAULT_RETRY)
    params = {'url': url, 'group': group}
    try:
        response = requests.post(croquemort_check_url,
                                 data=json.dumps(params),
                                 timeout=TIMEOUT)
    except requests.Timeout:
        raise UnreachableLinkChecker(TIMEOUT_LOG_MSG)
    except requests.RequestException as e:
        log.error(ERROR_LOG_MSG, exc_info=True)
        raise UnreachableLinkChecker('{}: {}'.format(ERROR_LOG_MSG, e))
    try:
        url_hash = response.json()['url-hash']
        retrieve_url = '{url}/url/{url_hash}'.format(
            url=croquemort_url, url_hash=url_hash)
    except ValueError:
        raise UnreachableLinkChecker('Wrong response for retrieve_url')
    response = None
    attempts = 0
    while is_pending(response):
        if attempts >= retry:
            msg = ('We were unable to retrieve the URL after'
                   ' {attempts} attempts.').format(attempts=attempts)
            raise UnreachableLinkChecker(msg)
        try:
            response = requests.get(retrieve_url, timeout=TIMEOUT)
        except requests.Timeout:
            raise UnreachableLinkChecker(TIMEOUT_LOG_MSG)
        except requests.RequestException:
            log.error(ERROR_LOG_MSG, exc_info=True)
            raise UnreachableLinkChecker('{}: {}'.format(ERROR_LOG_MSG, e))
        time.sleep(delay)
        attempts += 1
    return response.json()


udata_croquemort/checker.py
import logging

import dateutil.parser

from udata.models import Dataset
from .utils import check_url, UnreachableLinkChecker

log = logging.getLogger(__name__)


class CroquemortLinkChecker(object):
    """Croquemort link checker implementation.

    The main interface is the `check` method.
    """

    def _format_response(self, response):
        status = response.get('final-status-code')
        try:
            status = int(status)
        except ValueError:
            return {'check:error': 'Malformed check response'}
        else:
            check_date = response.get('updated')
            if check_date:
                check_date = dateutil.parser.parse(response.get('updated'))

            result = {
                'check:url': response.get('checked-url'),
                'check:status': status,
                'check:available': status and status >= 200 and status < 400,
                'check:date': check_date,
            }

            for header in [
                'content-type', 'content-length', 'content-md5', 'charset',
                'content-disposition'
            ]:
                value = response.get(header, '')
                if len(value) > 0:
                    try:
                        result[f"check:headers:{header}"] = int(value)
                    except ValueError:
                        result[f"check:headers:{header}"] = value

            return result

    def check(self, resource):
        """
        Parameters
        ----------
        resource : a uData resource instance to be checked

        Returns
        -------
        dict or None
            The formatted response from the linkchecker, like so:
            {
              'check:url': 'https://example.com',
              'check:status': 200,
              'check:available': True,
              'check:date': datetime.datetime(2017, 9, 4, 11, 13, 8, 888288),
              'check:headers:content-type': 'text/csv',
              'check:headers:content-length': 245436,
              'check:headers:content-md5': 'acbd18db4cc2f85cedef654fccc4a4d8',
              'check:headers:charset': 'utf-8',
              'check:headers:content-disposition': 'inline'
            }
            Or in case of failure (in udata-croquemort, not croquemort):
            {
              'check:error': 'Something went terribly wrong.'
            }
            Or in case of failure in croquemort:
            None
        """
        log.debug('Checking resource w/ URL %s', resource.url)
        dataset = Dataset.objects(resources__id=resource.id).first()
        if not dataset:
            message = 'No dataset found for resource %s', resource.id
            log.error(message)
            return {'check:error': message}
        try:
            # do not check ftp(s) urls
            if resource.url and resource.url.lower().startswith('ftp'):
                return
            check_response = check_url(resource.url, group=dataset.slug)
            return self._format_response(check_response)
        except UnreachableLinkChecker as e:
            log.error('Unreachable croquemort for url %s: %s', resource.url, e)
            return
